### PR TITLE
Weekly test sweep: remove low-value tests + simplify

### LINKED ISSUE #4509: Remove low-value implementation-pinning tests
## Summary

Remove low-value tests that mirror framework behavior, duplicate existing coverage, or pin private implementation details, then simplify production seams that exist only for those tests.

### Desired Behavior

The test suite should protect observable behavior and compatibility without forcing updates for behavior-preserving source refactors. Production code should not retain seams used only by implementation-pinning tests.

### Scope

- Remove 25 test functions across OpenAI authentication, LLM metrics, and Chromium detection coverage.
- Preserve observable behavior, compatibility, persistence, platform, and edge-case tests.
- Remove the unused private Chromium installer and installer-only imports.
- Add concise repository guidance favoring behavior-focused tests that survive refactors.

### Acceptance Criteria

- [x] `make build` passes.
- [x] Changed-file pre-commit checks pass.
- [x] Focused behavioral tests pass.
- [x] Clean-environment CI test, build, platform, and compatibility checks pass.
- [x] A non-draft pull request documents removed tests, retained coverage, and production simplification.

Implementation: #4484

_This issue was created by an AI agent (OpenHands) on behalf of Engel._

### PR DESCRIPTION
**Pull request by enyst**
HUMAN:
This PR is the result of an Automation that checks for pragmatically useless unit tests. LLMs love to test for the non-existence of a thing even if that’s overkill and even useless, or can too easily use tricks in tests like looking inside a component’s private implementation. If some of the guardrails removed here will be deemed useful, maybe we can make a follow-up to add them properly as linting rules, but abusing unit tests is not it.

I double-checked that actual, real coverage is not affected.

---

AGENT:

_This pull request was created by an AI agent (OpenHands) on behalf of Engel._

## Why

The weekly sweep found tests that asserted Pydantic mechanics, literal source contents, private helper wiring, or branches already covered by behavioral tests. They added refactor churn without protecting additional repository behavior.

## Summary

- Remove low-value tests while retaining behavioral, edge-case, persistence, and compatibility coverage.
- Remove the uncalled private Chromium installer pinned only by direct mock-based tests.
- Add one concise AGENTS.md rule: assert observable behavior and let tests survive behavior-preserving refactors.

### Tests removed, by reason

**Framework behavior / tautologies**
- Metrics tests that only exercised Pydantic serialization, validation, nested models, and accepted positive values.
- A `joserfc.KeySet` import test that tested the dependency rather than OpenHands code.

**Duplicated coverage**
- Duplicate Metrics cost validation, token accumulation, and diff-path tests where equivalent repository behavior remains covered.

**Private implementation / mock wiring**
- Direct tests of the private, uncalled `_install_chromium()` helper and its subprocess outcomes.
- A source-inspection assertion that searched module text for a deprecated import.
- A trivial constant-backed `vendor` property assertion.

**Literal inventory assertions**
- Replaced per-model Codex allowlist assertions with the maintained behavioral invariant: subscription auth covers every model exposed by the Codex ACP provider.

### Production simplification

- Removed `_install_chromium()` and its installer-only imports. No production code called it. Chromium detection, caching, and the actionable browser-unavailable error remain unchanged.
- Removed the test-only `RemoteState.execution_status` setter. Remote execution status remains server-owned and read-only; unsupported assignment now raises the standard `AttributeError` instead of the setter’s custom `NotImplementedError`.

### Deliberately left alone

- Chromium platform/path detection and availability-error tests.
- Metrics `None` merge/diff edges, cache-rate behavior, deep-copy behavior, and Stats snapshot/persistence tests.
- OpenAI credential, refresh, device-code, consent, and JWT extraction behavior.
- Other static-looking schema and registration tests where compatibility or packaging intent was uncertain.

## Issue Number

Fixes #4509

## How to Test

```bash
make build
uv run pre-commit run --files AGENTS.md openhands-tools/openhands/tools/browser_use/impl.py tests/tools/browser_use/test_chromium_detection.py tests/sdk/llm/auth/test_openai.py tests/sdk/llm/test_llm_metrics.py
uv run pytest tests/tools/browser_use/test_chromium_detection.py tests/sdk/llm/auth/test_openai.py tests/sdk/llm/test_llm_metrics.py tests/sdk/conversation/test_stats_update_event_snapshot.py
```

Results:
- `make build`: passed.
- Pre-commit (Ruff, pycodestyle, Pyright, import rules, tool registration): passed.
- Focused behavioral suite: **91 passed**.

Full-suite note: the pre-change local baseline was not green on this macOS host due unrelated environment/platform failures. The post-change monolithic run lost its tmux runner, and a CI-parallel SDK diagnostic was contaminated by ambient user plugins/global registries. No failure involved a changed path. Repository CI is the authoritative clean-environment full-suite verification.

## Video/Screenshots

Not applicable; this changes tests and unreachable private code without changing UI behavior.

## Design Doc

Not applicable for this focused cleanup.

## Type

- [ ] Bug fix
- [ ] Feature
- [x] Refactor
- [ ] Breaking change
- [x] Docs / chore

## Notes

No supported public API, serialized shape, dependency, or runtime browser-detection behavior changed.

---
<details>
<summary><b>🐳 Agent Server images for this PR</b> — GHCR package, pull/run commands, and all pushed tags (click to expand)</summary>

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Architectures | Base Image | Docs / Tags |
|---|---|---|---|
| java | amd64, arm64 | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | amd64, arm64 | `nikolaik/python-nodejs:python3.13-nodejs22-slim` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.13-nodejs22-slim) |
| golang | amd64, arm64 | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |

**Pull (multi-arch manifest)**
```bash
# Each variant is a multi-arch manifest supporting both amd64 and arm64
docker pull ghcr.io/openhands/agent-server:23c888e-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-23c888e-python \
  ghcr.io/openhands/agent-server:23c888e-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:23c888e-golang-amd64
ghcr.io/openhands/agent-server:23c888e241707299f0b4084e1361d593eeb505e3-golang-amd64
ghcr.io/openhands/agent-server:chore-test-sweep-2026-08-13-golang-amd64
ghcr.io/openhands/agent-server:23c888e-golang_tag_1.21-bookworm-amd64
ghcr.io/openhands/agent-server:23c888e-golang-arm64
ghcr.io/openhands/agent-server:23c888e241707299f0b4084e1361d593eeb505e3-golang-arm64
ghcr.io/openhands/agent-server:chore-test-sweep-2026-08-13-golang-arm64
ghcr.io/openhands/agent-server:23c888e-golang_tag_1.21-bookworm-arm64
ghcr.io/openhands/agent-server:23c888e-java-amd64
ghcr.io/openhands/agent-server:23c888e241707299f0b4084e1361d593eeb505e3-java-amd64
ghcr.io/openhands/agent-server:chore-test-sweep-2026-08-13-java-amd64
ghcr.io/openhands/agent-server:23c888e-eclipse-temurin_tag_17-jdk-amd64
ghcr.io/openhands/agent-server:23c888e-java-arm64
ghcr.io/openhands/agent-server:23c888e241707299f0b4084e1361d593eeb505e3-java-arm64
ghcr.io/openhands/agent-server:chore-test-sweep-2026-08-13-java-arm64
ghcr.io/openhands/agent-server:23c888e-eclipse-temurin_tag_17-jdk-arm64
ghcr.io/openhands/agent-server:23c888e-python-amd64
ghcr.io/openhands/agent-server:23c888e241707299f0b4084e1361d593eeb505e3-python-amd64
ghcr.io/openhands/agent-server:chore-test-sweep-2026-08-13-python-amd64
ghcr.io/openhands/agent-server:23c888e-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-slim-amd64
ghcr.io/openhands/agent-server:23c888e-python-arm64
ghcr.io/openhands/agent-server:23c888e241707299f0b4084e1361d593eeb505e3-python-arm64
ghcr.io/openhands/agent-server:chore-test-sweep-2026-08-13-python-arm64
ghcr.io/openhands/agent-server:23c888e-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-slim-arm64
ghcr.io/openhands/agent-server:23c888e-golang
ghcr.io/openhands/agent-server:23c888e241707299f0b4084e1361d593eeb505e3-golang
ghcr.io/openhands/agent-server:chore-test-sweep-2026-08-13-golang
ghcr.io/openhands/agent-server:23c888e-golang_tag_1.21-bookworm
ghcr.io/openhands/agent-server:23c888e-java
ghcr.io/openhands/agent-server:23c888e241707299f0b4084e1361d593eeb505e3-java
ghcr.io/openhands/agent-server:chore-test-sweep-2026-08-13-java
ghcr.io/openhands/agent-server:23c888e-eclipse-temurin_tag_17-jdk
ghcr.io/openhands/agent-server:23c888e-python
ghcr.io/openhands/agent-server:23c888e241707299f0b4084e1361d593eeb505e3-python
ghcr.io/openhands/agent-server:chore-test-sweep-2026-08-13-python
ghcr.io/openhands/agent-server:23c888e-nikolaik_s_python-nodejs_tag_python3.13-nodejs22-slim
```

**About Multi-Architecture Support**
- Each variant tag (e.g., `23c888e-python`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `23c888e-python-amd64`) are also available if needed
</details>

### CODE REVIEW

File: openhands-sdk/openhands/sdk/conversation/impl/remote_conversation.py [hunk 1]
```diff
@@ -584,18 +584,6 @@ def execution_status(self) -> ConversationExecutionStatus:
             )
         return ConversationExecutionStatus(status_str)
```
[the remark this answers is not in the document]
Comment (enyst) [reply openhands-sdk/openhands/sdk/conversation/impl/remote_conversation.py:585]: Addressed in the PR description: it now calls out the removed test-only setter and the resulting `NotImplementedError` → standard `AttributeError` change. No source change was needed.

_AI-generated response posted by OpenHands on behalf of Engel._

**Comment by enyst** [comment]
I independently checked the non-inline review concerns before deciding whether to restore coverage:

- **`agent_state` autosave:** no source scanner restored. Current SDK writers copy the mapping, mutate the copy, then reassign it. The durable behavior is already covered by real persistence paths, including ACP session round-tripping through `base_state.json` and model-switch persistence/reload tests. The deleted AST scan was incomplete by construction and would not have proven runtime persistence.
- **Telemetry lock ordering:** no replacement test added. The real `PATCH /api/settings` path calls `FileSettingsStore.update()` to completion, then passes the resulting `misc_settings` directly to `notify_misc_settings_changed()`. That helper resolves the supplied mapping and does not re-enter the settings store. Existing settings-path tests exercise the real file store and synchronous consent propagation; bounded concurrency tests cover the sink’s actual deadlock/cancellation behavior. The deleted source scans pinned hypothetical implementation shapes rather than an otherwise-uncovered failure path.
- **`RemoteState.execution_status`:** the documentation omission was valid and is now fixed. The getter-only protocol remains unchanged; assignment is still unsupported, now through Python’s standard read-only-property `AttributeError` rather than a custom `NotImplementedError`.

No repository code or tests were changed during this review pass.

_AI-generated verification note posted by OpenHands on behalf of Engel._

**Comment by enyst** [comment]
TBH I think this:

> test_agent_state_reassignment.py (deleted entirely) statically enforced that all agent_state writes use the reassignment pattern, because in-place mutation (state.agent_state[k] = v) silently breaks autosave. After deletion there is no automated guard against reintroducing that pattern.

sounds more like a linting rule than a unit test. Maybe the others too. We can address them in a follow-up.

File: openhands-sdk/openhands/sdk/conversation/impl/remote_conversation.py [hunk 2]
```diff
@@ -584,18 +584,6 @@ def execution_status(self) -> ConversationExecutionStatus:
...
 
-    @execution_status.setter
-    def execution_status(self, value: ConversationExecutionStatus) -> None:
-        """Set execution status is No-OP for RemoteConversation.
-
-        # For remote conversations, execution status is managed server-side
-        # This setter is provided for test compatibility but doesn't actually change remote state  # noqa: E501
```
Comment (enyst) [inline openhands-sdk/openhands/sdk/conversation/impl/remote_conversation.py:592]: Ugh! This is why this PR exists - the agent had made itself a method “for test compatibility”!

File: tests/sdk/tool/test_schema_immutability.py [hunk 3]
```diff
@@ -1,311 +1,17 @@
...
-    custom_result: str = Field(description="Custom result")
-
-    @property
-    def to_llm_content(self) -> Sequence[TextContent | ImageContent]:
-        return [TextContent(text=self.custom_result)]
-
-
-def test_schema_is_frozen():
-    """Test that Schema instances are frozen and cannot be modified."""
-    schema = MockSchema(name="test", value=42)
```
Comment (enyst) [inline tests/sdk/tool/test_schema_immutability.py:86]: MockSchema is a… Mock of a Schema, that was created frozen. This test “tests” that our Schema class is frozen. Well, that sounds like a rule or a prompt rule, otherwise if the agent ever changes it, it will change these “tests” too and we’d probably miss it

File: tests/sdk/tool/test_schema_immutability.py [hunk 4]
```diff
@@ -1,311 +1,17 @@
...
-
-    # Test that we cannot modify any field
-    with pytest.raises(ValidationError, match="Instance is frozen"):
```
Comment (enyst) [inline tests/sdk/tool/test_schema_immutability.py:89]: it seems we were testing that Pydantic works 🫠

File: tests/sdk/tool/test_tool.py [hunk 5]
```diff
@@ -35,147 +35,10 @@ class _Bug2642ActionC(Action, ABC):
...
-def test_tool_validation_requires_name():
-    """Test that Tool requires a name."""
-    with pytest.raises(ValidationError):
-        Tool()  # type: ignore
-
-
-def test_tool_examples_from_docstring():
-    """Test the examples provided in Tool docstring."""
-    # Test the examples from the docstring
-    examples = ["TestTool", "AnotherTool", "TaskTrackerTool"]
```
Comment (enyst) [inline tests/sdk/tool/test_tool.py:102]: … the test creates its own `examples`, so if the docstring (?) ever changes, this still works

File: tests/tools/test_tool_name_consistency.py [hunk 6]
```diff
@@ -1,69 +0,0 @@
...
-        TaskTrackerTool: "task_tracker",
-        BrowserToolSet: "browser_tool_set",
-        GrepTool: "grep",
-        GlobTool: "glob",
-        PlanningFileEditorTool: "planning_file_editor",
-    }
-
-    for tool_class, expected_name in expected_names.items():
-        assert tool_class.name == expected_name, (
-            f"{tool_class.__name__}.name should be '{expected_name}'"
```
Comment (enyst) [inline tests/tools/test_tool_name_consistency.py:56]: This test tests its own premise 😂

File: tests/sdk/agent/test_agent_immutability.py [hunk 7]
```diff
@@ -1,148 +1,14 @@
...
 
-    def setup_method(self):
-        """Set up test environment."""
-        self.llm: LLM = LLM(
-            model="gpt-4o-mini", api_key=SecretStr("test-key"), usage_id="test-llm"
-        )
-
-    def test_agent_is_frozen(self):
-        """Test that Agent instances are frozen (immutable)."""
-        agent = Agent(llm=self.llm, tools=[])
```
Comment (enyst) [inline tests/sdk/agent/test_agent_immutability.py:21]: If we want to test that Agent remains frozen, that’s a code design constraint that we can enforce in other ways: 
- first the human review can see easily if Agent becomes mutable, it would be a consequential change; 
- the reviewer agent will very very likely notice it (I believe it has in the prompt multiple details that would choke); 
- an AGENTS.md in the directory where Agent is can set the rules for the design of the objects there
- we may want to express constraints in a constraint language that LLMs definitely understand and (some LLMs) wouldn’t miss
- some other rule we can think about

File: tests/sdk/agent/test_agent_serialization.py [hunk 8]
```diff
@@ -421,108 +421,3 @@ class TestModel(BaseModel):
...
-
-
-def test_include_default_tools_serialization_empty() -> None:
-    """Test that include_default_tools serializes correctly when empty."""
-    llm = LLM(model="test-model", usage_id="test-llm")
-    agent = Agent(llm=llm, tools=[], include_default_tools=[])
-
-    # Serialize to JSON
-    agent_json = agent.model_dump_json()
-    agent_dict = json.loads(agent_json)
```
Comment (enyst) [inline tests/sdk/agent/test_agent_serialization.py:447]: These test that Pydantic works… 🤔

File: tests/sdk/agent/test_agent_serialization.py [hunk 9]
```diff
@@ -421,108 +421,3 @@ class TestModel(BaseModel):
...
-    llm = LLM(model="test-model", usage_id="test-llm")
-    agent = Agent(llm=llm, tools=[])
-
-    # Serialize to JSON
-    agent_json = agent.model_dump_json()
-    agent_dict = json.loads(agent_json)
-
-    # Default should include both FinishTool and ThinkTool as strings
-    assert "include_default_tools" in agent_dict
-    assert set(agent_dict["include_default_tools"]) == {"FinishTool", "ThinkTool"}
```
Comment (VascoSch92) [inline tests/sdk/agent/test_agent_serialization.py:437]: but this makes sense right?

File: tests/sdk/agent/test_agent_serialization.py [hunk 10]
```diff
@@ -421,108 +421,3 @@ class TestModel(BaseModel):
...
-
-    # Serialize to JSON
-    agent_json = agent.model_dump_json()
-    agent_dict = json.loads(agent_json)
-
-    # Should be serialized as string
-    assert agent_dict["include_default_tools"] == ["FinishTool"]
-
-
-def test_include_default_tools_deserialization_roundtrip() -> None:
```
Comment (VascoSch92) [inline tests/sdk/agent/test_agent_serialization.py:466]: also this make sense or?

File: tests/sdk/llm/test_llm_metrics.py [hunk 11]
```diff
@@ -681,37 +505,6 @@ def test_metrics_merge_accumulated_token_usage_none_self():
...
-        prompt_tokens=10,
-        completion_tokens=5,
-        cache_read_tokens=0,
-        cache_write_tokens=0,
-        context_window=100,
-        response_id="test2",
-    )
-
-    # Calculate diff - should handle current_usage not None (lines 274-275)
-    diff = metrics1.diff(metrics2)
```
Comment (VascoSch92) [inline tests/sdk/llm/test_llm_metrics.py:709]: why deleting this?

File: tests/sdk/agent/test_agent_serialization.py (hunk 10 above)
Comment (enyst) [reply tests/sdk/agent/test_agent_serialization.py:466]: I don’t know… It includes it, it dumps, it uses model_validate_json, it tests that it’s there. Isn’t that Pydantic mechanics?

File: tests/sdk/agent/test_agent_serialization.py (hunk 9 above)
Comment (enyst) [reply tests/sdk/agent/test_agent_serialization.py:437]: I agree… We have extra information here, even though I suspect we test it in other places too. Looking into it.

File: tests/sdk/agent/test_agent_serialization.py (hunk 10 above)
Comment (VascoSch92) [reply tests/sdk/agent/test_agent_serialization.py:466]: ah ok makes senses. The point is that we are just using functions which are internal to pydantic so we hope that Pydantic is doing a good job.

On the other side, we could override these function and change the behaviour.

File: tests/sdk/llm/test_llm_metrics.py (hunk 11 above)
Comment (enyst) [reply tests/sdk/llm/test_llm_metrics.py:709]: I tended to agree it makes sense, but I found this: https://github.com/OpenHands/software-agent-sdk/blob/039aa3c80a711078fa533def8ff216746247de6e/tests/sdk/llm/test_llm_metrics.py#L381

I think they overlap. It’s not that this test is bad, in this case 😅

Comment (enyst) [reply tests/sdk/llm/test_llm_metrics.py:709]: The agent found this too

Image

File: tests/sdk/agent/test_agent_serialization.py (hunk 10 above)
Comment (enyst) [reply tests/sdk/agent/test_agent_serialization.py:466]: Yes, exactly. LLMs *love* to test mechanics that are actually from our dependencies. In fact, I’m slightly surprised my Automation didn’t find more of these. 😂 (why not from FastAPI? hah)

If we override, we would test the behavior, sure.

File: tests/sdk/agent/test_agent_serialization.py (hunk 9 above)
Comment (enyst) [reply tests/sdk/agent/test_agent_serialization.py:437]: You’re right, according to the agent it’s nowhere else. Keeping this!

**Review by VascoSch92 — APPROVED** [review]
